galley/pkg/crd/synchronizer.go

```diff
@@ -0,0 +1,364 @@
+			newCrd := new.(*apiext.CustomResourceDefinition)
+			oldCrd := old.(*apiext.CustomResourceDefinition)
+			if newCrd.ResourceVersion == oldCrd.ResourceVersion {
+				// Periodic resync will send update events for all known Deployments.
+				// Two different versions of the same Deployment will always have different RVs.
+				return
+			}
+			s.handleEvent(change.Update, new)
+		},
+		DeleteFunc: func(obj interface{}) { s.handleEvent(change.Delete, obj) },
+	})
+
+	// start CRD shared informer background process.
+	go s.informer.Run(s.stopCh)
+
+	// Wait for CRD cache sync.
+	if !cache.WaitForCacheSync(s.stopCh, s.informer.HasSynced) {
+		log.Warnf("Shutting down while waiting for CRD cache sync")
+	}
+	log.Debugf("Completed CRD cache sync and starting processing.")
+
+	go s.process()
+
+	return nil
+}
+
+// Stop the Synchronizer.
+func (s *Synchronizer) Stop() {
+	s.stateLock.Lock()
+	defer s.stateLock.Unlock()
+
+	if s.stopCh == nil {
+		log.Warnf("crd.Synchronizer: The synchronizer is already stopped.")
+		// Already stopped
+		return
+	}
+	log.Info("Stopping CRD Synchronizer")
+
+	s.queue.ShutDown()
+
+	close(s.stopCh)
+	s.stopCh = nil
+}
+
+// transform the incoming object into a change.Info and enqueue it in the work queue.
+func (s *Synchronizer) handleEvent(t change.Type, obj interface{}) {
+	var object metav1.Object
+	var ok bool
+	if object, ok = obj.(metav1.Object); !ok {
+		var tombstone cache.DeletedFinalStateUnknown
+		tombstone, ok = obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			log.Errorf("error decoding object, invalid type: %v", reflect.TypeOf(obj))
+			return
+		}
+		object, ok = tombstone.Obj.(metav1.Object)
+		if !ok {
+			log.Errorf("error decoding object tombstone, invalid type: %v", reflect.TypeOf(tombstone.Obj))
+			return
+		}
+		log.Infof("Recovered deleted object '%s' from tombstone", object.GetName())
+	}
+
+	info := &change.Info{
+		Type: t,
+		Name: object.GetName(),
+	}
+
+	s.queue.AddRateLimited(info)
+}
+
+// The main processing loop that takes out work from the queue and processes it in a panic-protected context.
+func (s *Synchronizer) process() {
+	for {
+		item, shutdown := s.queue.Get()
+		if shutdown {
+			log.Debugf("exiting the crd synchronizer process loop due to shutdown signal")
+			break
+		}
+		log.Debugf("incoming item for processing: %v", item)
+
+		if s.processQueueItem(item) {
+			s.queue.Forget(item)
+			log.Debugf("item processing complete successfully: %v", item)
+		}
+
+		s.queue.Done(item)
+		if s.eventHook != nil {
+			s.eventHook(item)
+		}
+		log.Debugf("marked item as done: %v", item)
+	}
+}
+
+// processQueueItem is a panic-protected processor of work items.
+func (s *Synchronizer) processQueueItem(item interface{}) (result bool) {
+	reachedEnd := false
+	defer func() {
+		r := recover()
+		if !reachedEnd {
+			log.Errorf("panic in crd.Synchronizer.processChange: %v", r)
+			// TODO: This will cause the panic to keep recurring in a rate-limited way. This will cause many log
```

Comment:
Under what conditions do we expect this to panic? AFAICT, this is only used by the mock/testing code to make it easier to jump up the stack.

Reply:
There is really no case we actually expect this to panic. However, this is the main processing loop and any panics here is undesirable? IIRC, a uncaught panic will cause the process to be torn down, and can potentially put Galley in a crash-loop.

Reply:
As discussed offline, remove the panic handler.